internal/controller/logger_manager.go

```diff
@@ -0,0 +1,236 @@
+// Copyright 2024 Illumio, Inc. All Rights Reserved.
+
+package controller
+
+import (
+	"io"
+	"os"
+	"sync"
+	"time"
+
+	pb "github.com/illumio/cloud-operator/api/illumio/cloud/k8scluster/v1"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"google.golang.org/grpc/connectivity"
+)
+
+const (
+	maxBufferSize = 100
+	flushInterval = 5 * time.Second
+)
+
+type ClientConnInterface interface {
+	GetState() connectivity.State
+	Close() error
+}
+
+// BufferedGrpcWriteSyncer is a custom zap writesync that writes to a grpc stream
+// In case stream is not connected it will buffer to memory
+type BufferedGrpcWriteSyncer struct {
+	client   pb.KubernetesInfoService_SendLogsClient
+	conn     ClientConnInterface
+	buffer   []*zapcore.Entry
+	mutex    sync.Mutex
+	done     chan struct{}
+	logger   *zap.SugaredLogger
+	logLevel zap.AtomicLevel
+	encoder  zapcore.Encoder
+}
+
+// NewBufferedGrpcWriteSyncer returns a new BufferedGrpcWriteSyncer
+func NewBufferedGrpcWriteSyncer(client pb.KubernetesInfoService_SendLogsClient, conn ClientConnInterface) *BufferedGrpcWriteSyncer {
+	bws := &BufferedGrpcWriteSyncer{
+		client: client,
+		conn:   conn,
+		buffer: make([]*zapcore.Entry, 0, maxBufferSize),
+		done:   make(chan struct{}),
+	}
+	go bws.run()
+	return bws
+}
+
+// Close flushes buffered log data into grpc stream if possible, and closes the connection.
+func (b *BufferedGrpcWriteSyncer) Close() error {
+	b.mutex.Lock()
+	defer b.mutex.Unlock()
+	b.flush()
+	// Close the channel if not already closed.
+	select {
+	case <-b.done:
+		// Already closed; do nothing
+		return nil
+	default:
+	}
+	close(b.done)
+	return b.conn.Close()
+}
+
+// flush will attempt to dump buffer into GRPC stream if available
+func (b *BufferedGrpcWriteSyncer) flush() {
+	b.mutex.Lock()
+	defer b.mutex.Unlock()
+
+	if len(b.buffer) == 0 || b.conn == nil || b.conn.GetState() != connectivity.Ready {
+		return
+	}
+	lostLogs := false
+	lostLogEntries := 0
+
+	for _, logEntry := range b.buffer {
+		if err := b.sendLog(logEntry); err != nil {
+			b.logger.Error("Failed to send log entry",
+				zap.Error(err),
+				zap.Int("buffered_log_entries_count", len(b.buffer)),
+			)
```

Comment:
Remove this. This is emitting a log entry for every log entry that can't be sent.
The log at the bottom of the method is more useful, and less noisy.
```suggestion
```